File: 92 Reverse Linked List II/reverseLinkedListII.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        
        if( m == n )
            return head;
        
        int index = 1;
        ListNode cur = head;
        ListNode pre = null;
        ListNode next;
        
        while( cur != null ){
            if( index == m )
                break;
            pre = cur;
            cur = cur.next;
            index++;
        }
        ListNode head_point = pre;
        ListNode back_point = cur;
        
        pre = null;
        for(int i = 0; i < n-m+1; i++){
            next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
        
        if( head_point != null )
            head_point.next = pre;
        else
            head = pre;
        
        back_point.next = cur;
        
        return head;
    }
}

File: 92 Reverse Linked List II/reverseLinkedListII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if m == n:
            return head
        
        index = 1
        pre = None
        cur = head
        
        while cur != None:
            if index == m:
                break
            pre = cur
            cur = cur.next
            index += 1
            
        head_point = pre
        back_point = cur
        
        pre = None
        for i in range(0, n-m+1):
            next_one = cur.next
            cur.next = pre
            pre = cur
            cur = next_one
            
        if head_point != None:
            head_point.next = pre
        else:
            head = pre
            
        back_point.next = cur
        
        return head
        
        